File: Math/codeforces_1925B.py

# Problem link ===>> https://codeforces.com/problemset/problem/1925/B

import math

def gcd(a, b):
    if b == 0:
        return a
    return gcd(b, a % b)


def solve(x, n):

    # Calculate all possible divisors of x, handling potential overflow using math.sqrt and int():
    divisors = [i for i in range(1, int(math.sqrt(x)) + 1) if x % i == 0]
    divisors.extend([x // i for i in divisors if i * i != x])

    # Sort divisors in descending order:
    divisors.sort(reverse=True)

    # Find the first divisor that can be used to create n sub-problems:
    max_balance = 0
    for divisor in divisors:
        if n <= x // divisor:  # Check if n is less than or equal to x divided by divisor
            max_balance = divisor
            break

    return max_balance


def main():
    # Read the number of test cases:
    t = int(input())

    for _ in range(t):
        # Read the difficulty x and the number of sub-problems n:
        x, n = map(int, input().split())

        # Call the solve function and print the result:
        result = solve(x, n)
        print(result)


if __name__ == "__main__":
    main()

File: Math/codeforces_1925B.cpp

// Problem link ===>> https://codeforces.com/problemset/problem/1925/B
// submission link ===>> https://codeforces.com/contest/1925/submission/246683981

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int gcd(int a, int b)
{
    if (b == 0)
    {
        return a;
    }
    return gcd(b, a % b);
}

int main()
{
    int t;
    cin >> t;

    while (t--)
    {
        int x, n;
        cin >> x >> n;

        // Find all divisors of x
        vector<int> divisors;
        for (int i = 1; i * i <= x; ++i)
        {
            if (x % i == 0)
            {
                divisors.push_back(i);
                if (i * i != x)
                {
                    divisors.push_back(x / i);
                }
            }
        }

        // Sort divisors in descending order
        sort(divisors.rbegin(), divisors.rend());

        // Find the first divisor that can be used to create n sub-problems
        int max_balance = 0;
        for (int divisor : divisors)
        {
            if (n * divisor <= x)
            {
                max_balance = divisor;
                break;
            }
        }

        cout << max_balance << endl;
    }

    return 0;
}
